validation.go

```diff
@@ -28,21 +24,45 @@ func NewValidationResult(message string, fatal bool) ValidationResult {
 	}
 }
 
+// A collection of ValidationResults that implements set logic, meaning it holds unique entries only (no duplicates)
+type ValidationResults struct {
+	validationResults map[ValidationResult]struct{}
+}
+
+// NewValidationResults creates a new ValidationResults instance
+func NewValidationResults() ValidationResults {
+	return ValidationResults{
+		validationResults: map[ValidationResult]struct{}{},
+	}
+}
+
 // IsFatal returns true if there is one or more [ValidationResult] that is fatal.
-func (v ValidationResults) IsFatal() bool {
-	if len(v) == 0 {
+func (vrs *ValidationResults) IsFatal() bool {
+	if len(vrs.validationResults) == 0 {
 		return false
 	}
 
-	for _, vr := range v {
-		if vr.Fatal {
+	for k := range vrs.validationResults {
+		if k.Fatal {
 			return true
 		}
 	}
 
 	return false
 }
 
+// Add either adds the given [ValidationResult] to the set or does not if it already exists in the set
+func (vrs *ValidationResults) Add(vr ValidationResult) {
+	vrs.validationResults[vr] = struct{}{}
```

Comment:
for both `Add` and `Extend` we need to nil check the `validationResults` map. Consumers aren't forced to use the `NewValidationResults()` constructor, so we should check for the edge case of a literal instantiation.

Reply:
and then add a test for both methods trying to add/extend to a literally instantiation.

Reply:
Done, thanks!